proxy/coco/src/peer/run_state.rs

```diff
@@ -254,165 +275,77 @@ impl RunState {
         }
     }
 
-    fn handle_peer_event(event: PeerEvent) -> Vec<Command> {
-        match event {
-            PeerEvent::GossipFetch(FetchInfo {
-                result: PutResult::Applied(Gossip { urn, .. }),
-                ..
-            }) => vec![Command::Include(urn)],
-            PeerEvent::GossipFetch(_) => vec![],
-        }
-    }
-
     /// Handle [`input::Sync`]s.
     fn handle_peer_sync(&mut self, input: &input::Sync) -> Vec<Command> {
-        if let Status::Syncing { synced, syncs } = self.status {
+        if let Status::Syncing {
+            failed,
+            succeeded,
+            syncs,
+        } = &mut self.status
+        {
             match input {
-                input::Sync::Started(_peer_id) => {
-                    self.status = Status::Syncing {
-                        synced,
-                        syncs: syncs + 1,
-                    };
+                input::Sync::Started(peer_id) => {
+                    syncs.insert(*peer_id);
                 },
-                input::Sync::Failed(_peer_id) | input::Sync::Succeeded(_peer_id) => {
-                    self.status = if synced + 1 >= self.config.sync.max_peers {
-                        Status::Online {
-                            connected: self.connected_peers.len(),
-                        }
-                    } else {
-                        Status::Syncing {
-                            synced: synced + 1,
-                            syncs: syncs - 1,
-                        }
-                    };
+                input::Sync::Failed(peer_id) => {
+                    syncs.remove(peer_id);
+                    failed.insert(*peer_id);
+                },
+                input::Sync::Succeeded(peer_id) => {
+                    syncs.remove(peer_id);
+                    succeeded.insert(*peer_id);
                 },
             }
+
+            if failed.len() + succeeded.len() >= self.config.sync.max_peers {
+                self.status = Status::Online {
+                    connected: self.stats.connected_peers,
+                };
+            }
         }
 
         vec![]
     }
 
     /// Handle [`ProtocolEvent`]s.
     #[allow(clippy::wildcard_enum_match_arm)]
-    fn handle_protocol(&mut self, event: ProtocolEvent<Gossip>) -> Vec<Command> {
+    fn handle_protocol(&mut self, event: ProtocolEvent) -> Vec<Command> {
         match (&self.status, event) {
-            // Go from [`Status::Stopped`] to [`Status::Started`] once we are listening.
-            (Status::Stopped { .. }, ProtocolEvent::Listening(_addr)) => {
```

Comment:
`Listening` doesn't exist anymore and it's now `Endpoint::Up`, with `Endpoint::Down` being its inverse